### DISCUSSION TITLE
TCP UDP의 차이점을 말해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#215] [Network]
뭘까용

**Comment by Kimprodp** [comment]
1. TCP는 데이터를 전송하기 전에 데이터 전송을 위한 연결을 만드는 연결 지향 프로토콜이지만, UDP는 연결을 설정하지 않고 데이터를 개별 패킷으로 나누어 전송한다.
2. TCP는 전송된 데이터가 손실되거나 손상된 경우, TCP는 이를 감지하고 재전송하여 데이터의 신뢰성을 보장하지만, UDP는 손실되거나 손상된 패킷은 재전송하지 않기 때문에 데이터 전송의 신뢰성을 보장하지 않는다.
3. TCP는 데이터 패킷이 순서대로 도착하도록 보장하기 위해 순서가 뒤바뀐다면 순서를 재정렬 하지만 UDP는 순서를 보장해주지 않는다.
4. TCP에 비해 UDP는 단순하기 때문에 전송 속도가 더 빠르다.

**Comment by seunghye218** [comment]
## TCP란?

전송 제어 프로토콜(Transmission Control Protocol)은 네트워크를 통해 서로 다른 장치 간에 메시지를 교환하는 데 도움이 되는 통신용 연결 지향 프로토콜입니다. 컴퓨터 간에 데이터 패킷을 전송하는 기술을 확립하는 IP(인터넷 프로토콜)는 TCP와 함께 작동합니다.
TCP의 위치는 OSI 모델의 전송 계층에 있습니다. 또한 TCP는 발신자와 수신자 사이에 가상 연결을 설정하여 정보가 정확하게 전송되도록 보장합니다.

## UDP란?

UDP(User Datagram Protocol)는 전송 계층 프로토콜입니다. TCP와 달리 신뢰할 수 없고 연결이 없는 프로토콜입니다. 따라서 데이터를 전송하기 전에 연결을 설정할 필요가 없습니다. UDP는 네트워크를 통해 지연 시간이 짧고 손실 허용 연결을 설정하는 데 도움이 됩니다. UDP는 프로세스 간 통신을 가능하게 합니다.

---

공통점으로 TCP(전송 제어 프로토콜)와 UDP(사용자 데이터그램 프로토콜)는 모두 전송 계층의 프로토콜입니다.
UDP는 지연 없이 작업할 수 있기 때문에 비디오 스트리밍, 온라인 게임의 경우 더 좋습니다.
사진, 파일 전송, 웹 브라우징, 이메일 등과 같은 데이터를 전송하는 경우 TCP가 더 좋습니다. 데이터가 정확해야 전송되어야 하기 때문입니다.

## TCP vs UDP
  | TCP(전송 제어 프로토콜) | UDP(사용자 데이터그램 프로토콜)
-- | -- | --
서비스 유형 | TCP는 연결 지향 프로토콜입니다. |  연결 방향은 통신 장치가 데이터를 전송하기 전에 연결을 설정해야 하고 데이터 전송 후에 연결을 닫아야 함을 의미합니다. | UDP는 데이터그램 지향 프로토콜입니다. |  연결을 열거나, 연결을 유지하거나, 연결을 종료하는 데 드는 오버헤드가 없기 때문입니다. UDP는 브로드캐스트 및 멀티캐스트 유형의 네트워크 전송에 효율적입니다. |   |  
신뢰성 | TCP는 대상 라우터로의 데이터 전달을 보장하므로 안정적입니다. | UDP에서는 대상으로의 데이터 전달을 보장할 수 없습니다. 오류 검사 메커니즘 | TCP는 광범위한 오류 검사 메커니즘을 제공합니다. |   이는 흐름 제어 및 데이터 승인을 제공하기 때문입니다. | UDP에는 체크섬을 사용하는 기본적인 오류 확인 메커니즘만 있습니다. |  
승인(Acknowledgment) | 승인 세그먼트가 존재합니다. | 승인 세그먼트가 없습니다.
순서 | 데이터 순서화는 전송 제어의 기능입니다. |  프로토콜(TCP). 이는 패킷이 수신자에게 순서대로 도착한다는 것을 의미합니다. | UDP에는 데이터 순서가 없습니다. 순서가 필요한 경우 애플리케이션 계층에서 관리해야 합니다. |  
속도 | TCP는 UDP보다 비교적 느립니다. | UDP는 TCP보다 빠르고 간단하며 효율적입니다.
재전송 | 손실된 패킷의 재전송은 TCP에서는 가능하지만 UDP에서는 불가능합니다. | UDP(사용자 데이터그램 프로토콜)에서는 손실된 패킷을 재전송하지 않습니다.
헤더 길이 | TCP에는 (20-60)바이트의 가변 길이 헤더가 있습니다. | UDP에는 8바이트의 고정 길이 헤더가 있습니다.
무게 | TCP는 무겁습니다. | UDP는 가볍습니다.
핸드쉐이킹 기법 | SYN, ACK, SYN-ACK와 같은 핸드셰이크를 사용합니다. | 연결이 없는 프로토콜입니다. 즉, 핸드셰이크가 없습니다.
Broadcasting | TCP는 브로드캐스팅을 지원하지 않습니다. | UDP는 브로드캐스팅을 지원합니다.
프로토콜 | HTTP, HTTPS, FTP, SMTP, Telnet  |  DNS, DHCP, TFTP, SNMP, RIP, VolP
스트림 유형 | TCP 연결은 바이트 스트림입니다. | UDP 연결은 메시지 스트림입니다.
Overhead | 낮지만 UDP보다 높습니다. | 매우 낮습니다.
Applications | 이 프로토콜은 이메일, 웹서핑, 군 복무 등 안전하고 신뢰할 수 있는 통신 절차가 필요한 상황에서 주로 활용됩니다. | 이 프로토콜은 VoIP, 게임 스트리밍, 비디오, 음악 스트리밍 등 빠른 통신이 필요하지만 신뢰성이 문제가 되지 않는 상황에서 사용됩니다.

**Comment by J-I-H-O** [comment]
TCP와 UDP는 전송 계층(Transport Layer)에서 사용하는 프로토콜입니다.

- **TCP (Transmission Control Protocol)**
    - **연결 지향 (connection-oriented)**
        - 데이터 전송 전에 송신자와 수신자 간 연결을 설정합니다.
        - 3-way handshaking 을 통해 연결을 설정하고, 4-way handshaking 을 통해 연결을 해제합니다.
    - **신뢰성**
        - 패킷 손실 시 재전송합니다.
        - 데이터의 순서를 보장합니다.
            - 패킷에 번호를 부여하고 목적지에서 재조립합니다. 이 과정에서 패킷의 분실 여부 및 순서를 확인합니다.
    - **흐름 제어 및 혼잡 제어 (flow control, congestion control)**
        - 흐름 제어를 통해 송신자가 데이터 전송 속도를 조절하여 수신자의 처리 능력을 초과하지 않도록 합니다. (buffer overflow 방지)
        - 혼잡 제어를 통해 네트워크 혼잡도를 최소화합니다.
    - **오버헤드**
        - 위와 같은 과정이 필요하기 때문에 UDP에 비해 속도가 느립니다.
        - 따라서 TCP는 연속성보다 신뢰성 있는 전송이 필요할 때에 사용됩니다.
            - 예) 파일, 이메일 전송 등
- **UDP (User Datagram Protocol)**
    - **비연결 지향 (connectionless)**
        - 데이터 전송 전에 연결을 설정하지 않습니다. 
        즉, 정보를 주고 받을 때 정보를 보내거나 받는다는 신호 절차를 거치지 않습니다.
        - 각 패킷은 독립적인 관계를 지니며, 다른 경로로 전송될 수 있습니다.
    - **신뢰성 없음**
        - 패킷 손실 시 재전송하지 않으며, 데이터 순서가 보장되지 않습니다.
        - UDP 헤더의 CheckSum 필드를 통해 최소한의 오류만 검출합니다.
    - **흐름 제어나 혼잡 제어 없음**
    - **오버헤드**
        - 패킷 순서 보장, 연결 수립, 흐름/혼잡 제어와 같은 과정을 수행하지 않기 때문에 빠릅니다.
        - 단, 신뢰성을 보장하지 못하기 때문에 신뢰성보다는 연속성이 중요한 서비스에 사용됩니다.
            - 예) 스트리밍 서비스, 온라인 게임 등

https://mangkyu.tistory.com/15
